dali/operators/reader/loader/webdataset_loader.cc

```diff
@@ -0,0 +1,371 @@
+#include "dali/util/file.h"
+
+namespace dali {
+
+template <typename... Args>
+inline std::string IndexFileErrMsg(const std::string& index_path, int64_t line,
+                                   const Args&... details) {
+  return make_string("Malformed index file at ", index_path, " line ", line, " - ", details...);
+}
+
+namespace detail {
+namespace wds {
+
+inline MissingExtBehavior ParseMissingExtBehavior(std::string missing_component_behavior) {
+  std::transform(missing_component_behavior.begin(), missing_component_behavior.end(),
+                 missing_component_behavior.begin(), static_cast<int (*)(int)>(std::tolower));
+  if (missing_component_behavior == "") {
+    return MissingExtBehavior::Empty;
+  } else if (missing_component_behavior == "skip") {
+    return MissingExtBehavior::Skip;
+  } else if (missing_component_behavior == "empty") {
+    return MissingExtBehavior::Empty;
+  } else if (missing_component_behavior == "error") {
+    return MissingExtBehavior::Raise;
+  } else {
+    return MissingExtBehavior::Invalid;
+  }
+}
+
+
+inline void ParseSampleDesc(std::vector<SampleDesc>& samples_container,
+                            std::vector<ComponentDesc>& components_container,
+                            std::ifstream& index_file, const std::string& index_path,
+                            int64_t line) {
+  // Preparing the SampleDesc
+  samples_container.emplace_back();
+  samples_container.back().components =
+      VectorRange<ComponentDesc>(components_container, components_container.size());
+
+  // Getting the components data
+  std::string components_metadata;
+  std::getline(index_file, components_metadata);
+  std::stringstream extensions_stream(components_metadata);
+
+  // Reading consecutive components
+  ComponentDesc component;
+  while (extensions_stream >> component.ext) {
+    DALI_ENFORCE(extensions_stream >> component.offset >> component.size,
+                 IndexFileErrMsg(index_path, line,
+                                 "size or offset corresponding to the extension not found"));
+    DALI_ENFORCE(
+        component.offset % kBlockSize == 0,
+        IndexFileErrMsg(index_path, line, "tar offset is not a multiple of tar block size (",
+                        kBlockSize, "), perhaps the size value is exported before offset?"));
+    components_container.emplace_back(std::move(component));
+    samples_container.back().components.num++;
+  }
+
+  // Finishing up the SampleDesc
+  DALI_ENFORCE(samples_container.back().components.num,
+               IndexFileErrMsg(index_path, line, "no extensions provided for the sample"));
+}
+
+inline void ParseIndexFile(std::vector<SampleDesc>& samples_container,
+                           std::vector<ComponentDesc>& components_container,
+                           const std::string& index_path) {
+  std::ifstream index_file(index_path);
+
+  // Index Checking
+  std::string index_version;
+  DALI_ENFORCE(index_file >> index_version,
```

Comment:
Read that using `getline()` - otherwise your line numbering will be destroyed at this point.

Reply:
alright